### PR TITLE
Split book entity 198

### LINKED ISSUE #198: Split Book entity
**Context**

The `Book` entity currently includes the following fields:

>  // For books that have been read
    private RatingScale rating;
    private LocalDate dateStartedReading;
    private LocalDate dateFinishedReading;
    private String bookReview;

The comment indicates that those four fields are only applicable to books that have been read, and not just any book (e.g. a book that is yet to be read).

**What needs to be done**

The `Book` entity could be split into a new class that inherits from `Book` called `ReadBook` with those fields.

### PR DESCRIPTION
**Pull request by Hassan-Elseoudy**
## Summary of change

So main change that I added a new feature, Book Progress .. A New table that connects between Users and Book.
Book Progress has a state of 3 (To be read, In Progress and Completed), A book progress will be defined by state and other parameters like Start Date, End Date, number of pages read, my ratings and my review.

I tried to decouple the book from its life cycle that is related to the user. 
Also added some tests, and a Minor change for `MAX_PAGE` to be handled with `@Range` annotation rather than `@Max`.

Also, please note that `pages` for each book is `Required` for the book progress, A system can track User's progress in reading book by checking user's read pages and book's pages.

I left all of those stuff dependent, I didn't change anything till you decide to inject those classes in your actual services.

## Related issue

Closes #198

## Pull request checklist

Please keep this checklist in & ensure you have done the following:

- [x] Read, understood and adhered to our [contributing document](https://github.com/knjk04/book-project/blob/master/CONTRIBUTING.md).
  - [x] Ensure that you were first assigned to a relevant issue before creating this pull request
  - [x] Ensure code changes pass all tests

- [x] Read, understood and adhered to our [style guide](https://github.com/knjk04/book-project/blob/master/STYLEGUIDE.md). A lot of our code reviews are spent on ensuring compliance with our style guide, so it would save a lot of time if this was adhered to from the outset.  (**I may be missing this, sorry :)** )

- [x] Filled in the summary, context (if applicable) and related issue section. Replace the square brackets and its placeholder content with your contents. For an example, see any merged in pull request
  - [x] Included a screenshot(s) if a UI change was involved (it may look different on a reviewer's device)

- [x] Created a branch that has a descriptive name (what your branch is for in a few words and includes the issue number at the end, e.g. `test-reading-goal-123`

- [x] Set this pull request to 'draft' if you are still working on it

- [x] Resolved any merge conflicts

For any of the optional checkboxes (e.g. the screenshots one), still check it if it does not apply.

If in doubt, get in touch with us via our [Slack workspace](https://join.slack.com/t/teambookproject/shared_invite/zt-i7lept44-rgJ9yB0A2vedJTLyyfkjKQ)

### CODE REVIEW

**Review by knjk04 — COMMENTED** [review]
Thanks for working on this!

```java
// For books that have been read
private RatingScale rating;
private LocalDate dateStartedReading;
private LocalDate dateFinishedReading;
private String bookReview;
```

From a quick look, unfortunately, this seems to add another layer of complexity without providing much value. The above code snippet is from the `Book` class, which was also in the issue description. The `Book` class still has lots of optional fields, and we may add more later (we've already added a couple of new fields recently). This solution doesn't not solve this problem, unfortunately.

I still like the idea of having different states that a book could be in, but I am going to look into this further to see if I can come up with a good alternative. 

```java
     void shouldntRateUnlessFinishTestWithExceptionThrown() throws Exception {
```

Also, just a tip -- throwing the `Exception` class is not a good idea (better to throw a narrow exception if we must). In the above code snippet, you've added a `throws` declaration. We don't want to catch Exception, as this may also catch system exceptions (e.g. out of memory) that we shouldn't be catching in the first place (we may catch something like out of memory that we weren't intending to)

**Comment by knjk04** [comment]
I appreciate the effort you put into writing this. However, we can't accept this, so I'll close this PR. I'll also unassign you from the issue. However, feel free to suggest a different approach. If I come up with an alternative implementation (which will be after Hacktoberfest), you can request to be assigned to the issue again.

File: src/main/java/com/karankumar/bookproject/backend/entity/Book.java [hunk 1]
```diff
@@ -101,8 +98,9 @@ public Book(String title, Author author, PredefinedShelf predefinedShelf) {
         this.title = title;
         this.author = author;
         this.predefinedShelf = predefinedShelf;
+        this.numberOfPages = 100;
```
[the remark this answers is not in the document]
Comment (Hassan-Elseoudy) [reply src/main/java/com/karankumar/bookproject/backend/entity/Book.java:101]: It was a default value, A Book progress is handled by checking how many pages you finished.
So to not break your work, I decided to leave it `100` for now, This should be given from client when he adds a new book.

File: src/main/java/com/karankumar/bookproject/backend/entity/book/BookProgress.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    @EmbeddedId
+    private BookProgressId id;
+
+    private RatingScale rating;
+
+    @JsonSerialize(using = LocalDateSerializer.class)
+    private LocalDate dateStartedReading;
+
+    @JsonSerialize(using = LocalDateSerializer.class)
+
```
[the remark this answers is not in the document]
Comment (Hassan-Elseoudy) [reply src/main/java/com/karankumar/bookproject/backend/entity/book/BookProgress.java:50]: `Book Progress` concept should store your start and finish times.

Benefits:
- You can track your reading rate.
- You can know which book you finished the fastest.
- You can reset your progress (Hint: User can save his latest start and finish dates)

So progress should have a `dateFinishedReading` field.

**Comment by knjk04** [comment]
Reopened at your request so that you can address the feedback

File: src/main/java/com/karankumar/bookproject/backend/entity/book/BookProgress.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (knjk04) [reply src/main/java/com/karankumar/bookproject/backend/entity/book/BookProgress.java:50]: You have:

```java
@JsonSerialize(using = LocalDateSerializer.class)    
// blank line here to be deleted
private LocalDate dateFinishedReading;
```

File: src/main/java/com/karankumar/bookproject/backend/entity/book/BookProgress.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+import java.time.LocalDate;
+
+@Entity
+@AllArgsConstructor
+@Getter
+@Setter
+@NoArgsConstructor
+public class BookProgress {
+
+    @EmbeddedId
```
[the remark this answers is not in the document]
Comment (Hassan-Elseoudy) [reply src/main/java/com/karankumar/bookproject/backend/entity/book/BookProgress.java:41]: As far as I know, a Book progress is defined by `user` reads a `book`.
So our primary key is a composite id of `User Id` and `Book Id`.

**Comment by knjk04** [comment]
Reopened at your request so that you can address the feedback

I've resolved some of the conversations as they were from an unfinished review.

**Comment by Hassan-Elseoudy** [comment]
> Reopened at your request so that you can address the feedback
> 
> I've resolved some of the conversations as they were from an unfinished review.

Thank you for opening this again. I'm already solving some of your comments :)

**Comment by knjk04** [comment]
>Thank you for opening this again. I'm already solving some of your comments :)

That's fine, but the design will also need to be changed, so we can discuss this over the help channel on Slack

**Comment by Hassan-Elseoudy** [comment]
> Thanks for working on this!
> 
> ```java
> // For books that have been read
> private RatingScale rating;
> private LocalDate dateStartedReading;
> private LocalDate dateFinishedReading;
> private String bookReview;
> ```
> 
> From a quick look, unfortunately, this seems to add another layer of complexity without providing much value. The above code snippet is from the `Book` class, which was also in the issue description. The `Book` class still has lots of optional fields, and we may add more later (we've already added a couple of new fields recently). This solution doesn't not solve this problem, unfortunately.
> 

I disagree with you, surely It will
1. We segregate the book it self and user's progress. By this, We are more comfortable to move non-related attributes of book to Book Progress. And this builds up to another version later by using `states` as an `enum` rather than a class for now. (It can be enhanced later by saving the current state and using Finite State Machine.

2. I'm pretty sure that if you have that much of optional fields, you have to switch to NoSQL, Although from my side, I guess there're less optional fields that you think.

> I still like the idea of having different states that a book could be in, but I am going to look into this further to see if I can come up with a good alternative.
> 
> ```java
>      void shouldntRateUnlessFinishTestWithExceptionThrown() throws Exception {
> ```
> 
> Also, just a tip -- throwing the `Exception` class is not a good idea (better to throw a narrow exception if we must). In the above code snippet, you've added a `throws` declaration. We don't want to catch Exception, as this may also catch system exceptions (e.g. out of memory) that we shouldn't be catching in the first place (we may catch something like out of memory that we weren't intending to)

Improved with better exceptions methods, And We can work on a Mechanism to handle errors and exceptions (Already build in spring boot) But I'd prefer working on it in another issue.

**Comment by knjk04** [comment]
In `withPagesRead()` in `BookProgressBuilder.java`, you're using the page number to assess state a book should be in. Currently, we're only displaying the pages read field in the book form if the book was in the 'did not finish' state. The way we determine which state a book is currently in is by letting a user choose in the book form: they can easily move a book to any shelf they choose. We don't want to lose this flexibility. 

Furthermore, it seems possible to completely forego the progress service layer and opt to use the simpler approach of using the `BookService` as is. 

---

I still like the idea of states, but I think there's an easier way to solve this problem. I don't yet know what this is, so I'll take some to think about this. In the meantime, I'll block the issue again and unassign you. You're still free to come up with alternative solutions and let us know (via Slack). If we can agree on something that's best for the project, then we can go with it, and I can re-assign you. 

Unfortunately, I stand by what I said yesterday about this PR not fitting what we're looking for right now. I'll close this pull request. If you'd like to respond to my feedback, you can still do so on Slack on the thread you made on Slack. In line with our contributing guidelines, it's better to have an open-ended discussion there. 

I appreciate this may be frustrating and I can only hope that you're not deterred from contributing again.

**Pull request closed without merging**